### PR TITLE
Well... lots of stuff, see comment

### PR DESCRIPTION
**Pull request by lociii**
- Completely restructured code in small functions
- Add lots of error handling when BLE stuff fails
- Support multiple devices
- Add device retry
- Change how deep sleep is activated (no more unwanted sleep while devices are still updated)
- Add emergency deep sleep handler in case something gets stuck (should use a quite long delay)
- Make log output easier to read
- Properly disconnect services

Guess that's all :-)

The reboot issues noted in #9 and #10 are handled with these updates.

### CODE REVIEW

**Comment by sidddy** [comment]
Wow, that's great!! Lots of stuff in there which I had planned to do, but never found time... I'll review the changes soon!

File: flora/config.h.example [hunk 1]
```diff
@@ -1,20 +1,31 @@
...
 
-#define MQTT_TEMPERATURE MQTT_DEV_BASE "temperature"
-#define MQTT_MOISTURE MQTT_DEV_BASE "moisture"
-#define MQTT_LIGHT MQTT_DEV_BASE "light"
-#define MQTT_CONDUCTIVITY MQTT_DEV_BASE "conductivity"
-#define MQTT_BATTERY MQTT_DEV_BASE "battery"
+// sleep between to runs in seconds
+#define SLEEP_DURATION 30 * 60
+// emergency hibernate countdown in seconds
+#define EMERGENCY_HIBERNATE 3 * 60 * 1000
```
Comment (sidddy) [inline flora/config.h.example]: Comment says "seconds", but you specify milliseconds here. CHange define to 
```suggestion
#define EMERGENCY_HIBERNATE 3 * 60
```
and also change delay coding below.

File: flora/flora.ino [hunk 2]
```diff
@@ -100,123 +189,190 @@ bool getSensorData(BLEAddress pAddress, bool getBattery) {
...
+
+void hibernate() {
   esp_sleep_enable_timer_wakeup(SLEEP_DURATION * 1000000ll);
   Serial.println("Going to sleep now.");
+  delay(100);
   esp_deep_sleep_start();
 }
 
+void delayedHibernate(void *parameter) {
+  delay(EMERGENCY_HIBERNATE); // delay for five minutes
```
Comment (sidddy) [inline flora/flora.ino]: please change the coding here, to use seconds for EMERGENCY_HIBERNATE

```suggestion
  delay(EMERGENCY_HIBERNATE*1000); // delay for five minutes
```

File: flora/flora.ino [hunk 3]
```diff
@@ -100,123 +189,190 @@ bool getSensorData(BLEAddress pAddress, bool getBattery) {
...
-      Serial.println("connected");
-    } else {
-      Serial.print("failed, rc=");
-      Serial.print(client.state());
-      Serial.println(" try again in 5 seconds");
-      // Wait 5 seconds before retrying
-      delay(5000);
+  Serial.println("Initialize BLE client...");
+  BLEDevice::init("");
+  BLEDevice::setPower(ESP_PWR_LVL_P7);
```
Comment (sidddy) [inline flora/flora.ino]: uh, nice, didn't know that this feature exists :-)
does it actually help for better connectivity to the flora sensors?

File: flora/flora.ino [hunk 4]
```diff
@@ -100,123 +189,190 @@ bool getSensorData(BLEAddress pAddress, bool getBattery) {
...
-  Serial.print("Light: ");
-  Serial.println(light);
-  snprintf(buffer, 64, "%d", light);
-  client.publish(MQTT_LIGHT, buffer);
+  char buffer[64];
 
-  Serial.print("Conductivity: ");
-  Serial.println(conductivity);
+  snprintf(buffer, 64, "%f", temperature);
+  client.publish((baseTopic + "temperature").c_str(), buffer);
```
Comment (sidddy) [inline flora/flora.ino]: we could carve out these formatting & publishing instructions into a separate function... it looks too repetitive :-) Anyway, there's no real need to do this in this PR.

File: flora/flora.ino (hunk 3 above)
Comment (lociii) [reply flora/flora.ino]: Cannot really tell since it doesn't connect reliably anyways. :-)

File: flora/flora.ino (hunk 4 above)
Comment (lociii) [reply flora/flora.ino]: :+1:

**Pull request merged**